### PR TITLE
Add CI for docker builds

### PR DESCRIPTION
**Pull request by eastata**
Issue #278 

# Added
CI will:
* Build and push to registry docker images `zeth-mpc`, `zeth-client`, `zeth-prover` with docker tag based on git commit [Short SHA-1](https://git-scm.com/book/en/v2/Git-Tools-Revision-Selection)
  * For every commit to **develop** branch
  * For every PR to **master** branch 
* Build and push to registry docker images `zeth-mpc`, `zeth-client`, `zeth-prover` with docker tag == git tag and update `latest` images
  * For every git **tag** matched with [Semantic Versioning 2.0.0](https://semver.org/) like `vX.Y.Z`
* Build and push to registry docker image `zeth-base` with docker tag version from git tag and update `latest`
  * For every git **tag** matched with [Semantic Versioning 2.0.0](https://semver.org/) like `docker-base-vX.Y.Z`

# Updated
* Current `clearmatics/zeth-base:latest` image saved as a `v0.0.0` in DockerHub DIGEST:sha256:0a3c603a607e970e093b7a384f92e5958bba2dbb0af0eb1eaf786f8fc41d3381
* Source image for `zeth-mpc`, `zeth-dev`, `zeth-prover` moved from  `clearmatics/zeth-base:latest` to fixed version  `clearmatics/zeth-base:v0.0.0`

# Removed
* DockerHub Autobuild for `clearmatics/zeth-base` disabled

### CODE REVIEW

File: Dockerfile-dev [hunk 1]
```diff
@@ -1,4 +1,4 @@
-FROM clearmatics/zeth-base:latest
+FROM clearmatics/zeth-base:v0.0.0
```
Comment (AntoineRondelet) [inline Dockerfile-dev:1]: Why did you switch to `v0.0.0`? We want to always build the `dev` and subsequent images on top of the `zeth-base:latest` image on master.

File: README.md [hunk 2]
```diff
@@ -165,6 +163,16 @@ cmake -DCMAKE_C_COMPILER=/usr/bin/clang -DCMAKE_CXX_COMPILER=/usr/bin/clang++ -D
 make check
 ```
 
+## Docker images
+| Docker files | Image | Tags | Description |
+|---------------|------|-----|--|
+| [./Dockerfile-prover](./Dockerfile-prover) | [clearmatics/zeth-prover](https://hub.docker.com/r/clearmatics/zeth-prover) | `latest`, `vX.Y.Z` - Release of zeth, `git-%HASH%` - developers build by git-commit  | [Zeth Prover Server](./prover_server/README.md). Image use `zeth-base` for building |
+| [./Dockerfile-client](./Dockerfile-client) | [clearmatics/zeth-client](https://hub.docker.com/r/clearmatics/zeth-client) | `latest`, `vX.Y.Z` - Release of zeth, `git-%HASH%` - developers build by git-commit  | [Python client to interact with the prover](./client/README.md) |
+| [./Dockerfile-mpc](./Dockerfile-mpc) | [clearmatics/zeth-mpc](https://hub.docker.com/r/clearmatics/zeth-mpc) | `latest`, `vX.Y.Z` - Release of zeth, `git-%HASH%` - developers build by git-commit  | [Tools for Multi-Party Computation](./mpc/README.md). Image use `zeth-base` for building |
+| [./Dockerfile-base](./Dockerfile-base) | [clearmatics/zeth-base](https://hub.docker.com/r/clearmatics/zeth-base) | `latest`, `vA.B.C` - Release of zeth-base | Base image for building other containers |
```
Comment (AntoineRondelet) [inline README.md:172]: `vA.B.C` -> `vX.Y.Z`? For consistency or did you use `A.B.C` and `X.Y.Z` explicitly to represent the fact that the versioning of `zeth-base` and of the rest of the project can be independent?

**Comment by AntoineRondelet** [comment]
Great work, and thanks for the PR @eastata! That's very helpful!

File: .github/workflows/release-docker-base.yml [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    timeout-minutes: 360
+
+    steps:
+
+      - name: "Checkout"
+        uses: actions/checkout@v2
+
+      - name: "Get vars from git"
+        id: git_vars
+        run: echo ::set-output name=TAG::${GITHUB_REF/refs\/tags\/docker-base-/}
```
[the remark this answers is not in the document]
Comment (eastata) [reply .github/workflows/release-docker-base.yml:25]: yes

File: README.md (hunk 2 above)
Comment (eastata) [reply README.md:172]: Yes, zeth-base and zeth-apps is different products just placed in one repo. Release cycle is different.

Available options:
* Build all artefacts as one product sequently for each commit (it takes ages) or
* Split the products releases

File: Dockerfile-dev (hunk 1 above)
Comment (eastata) [reply Dockerfile-dev:1]: I would like to recommend you avoid using tag `latest` by the reasons:
* local docker, kubernetes nodes, compose etc cache images if the tags the same. 
* versions allow us to track source of images

If this points not applicable for `dev` image building I can roll it back to tag `latest`

Should I do it?

Comment (eastata) [reply Dockerfile-dev]: reverted

File: README.md (hunk 2 above)
Comment (AntoineRondelet) [reply README.md:172]: Thanks, makes sense

File: Dockerfile-dev (hunk 1 above)
Comment (AntoineRondelet) [reply Dockerfile-dev]: Yes, it's handy to build from `latest` for the `dev` image (we build the dev image on some of our tests on the CI server where we fetch the latest image everytime the CI runs). Thanks for reverting.

**Comment by AntoineRondelet** [comment]
@eastata can you rebase your branch on top of develop (that'll bring the recent changes that fix the CI for macOS and repair the CI on this PR - see here: https://github.com/clearmatics/zeth/pull/316).

Also, I saw that you added a tag: https://github.com/clearmatics/zeth/releases/tag/docker-base-v0.0.1
Is this one needed? or was it just for testing purposes? (seems like no changes are made to the Dockerfile-base file which would justify a version bump).

**Comment by eastata** [comment]
git tag `docker-base-v0.0.1` was to test build docker image `clearmatics/zeth-base:v0.0.1`

**Comment by AntoineRondelet** [comment]
> git tag `docker-base-v0.0.1` was to test build docker image `clearmatics/zeth-base:v0.0.1`

Alright, let's keep it as a "template" to follow for the next versions of `zeth-base`.
Note: I saw that you merged develop here, let's keep it that way for simplicity (and to merge this PR asap), but FYI the convention here is to *rebase* to avoid adding "bi-directional merge commits" in the git history (merging is kept unidirectional as much as we can i.e. `master` <- `develop` <- `feature` or `master` <- `hotfix`) :)

**Pull request merged**